src/tracked_struct.rs

```diff
@@ -524,56 +532,74 @@ where
         // Acquire the write-lock. This can only fail if there is a parallel thread
         // reading from this same `id`, which can only happen if the user has leaked it.
         // Tsk tsk.
-        let swapped_out = unsafe { (*data_raw).updated_at.swap(None) };
-        if swapped_out != last_updated_at {
+
+        // SAFETY: `updated_at` is never exclusively borrowed, so borrowing it is sound
+        let swapped = unsafe { (*data_raw).updated_at.swap(None) };
+        if last_updated_at != swapped {
             panic!(
                 "failed to acquire write lock, id `{id:?}` must have been leaked across threads"
             );
         }
 
-        // UNSAFE: Marking as mut requires exclusive access for the duration of
-        // the `mut`. We have now *claimed* this data by swapping in `None`,
-        // any attempt to read concurrently will panic.
-        let data = unsafe { &mut *data_raw };
+        // SAFETY: We have now *claimed* mutable access to the `value` field by swapping in `None`,
+        // any attempt to read concurrently will panic so it is safe to take exclusive references.
+        // FIXME(1.82.0): Replace `addr_of_mut!` with `&raw mut`
+        let old_fields =
+            unsafe { self.to_self_ptr(std::ptr::addr_of_mut!((*data_raw).value.fields)) };
+
+        // SAFETY: FIXME: why can this not alias with tracked_field::maybe_changed_after?
+        let revisions = unsafe { &mut (*data_raw).revisions };
 
         // SAFETY: We assert that the pointer to `data.revisions`
         // is a pointer into the database referencing a value
         // from a previous revision. As such, it continues to meet
         // its validity invariant and any owned content also continues
         // to meet its safety invariant.
+        let updated = unsafe { C::update_fields(current_revision, revisions, old_fields, fields) };
+
+        // SAFETY: The mutable accesses to the following metadata fields will not alias
+        // as attempting to read any of these would first verify whether we are up to date, either
+        // blocking on our update or implying we would not have ended up in this method here in the
+        // first place.
+        // FIXME: why can these not alias
+        // - `durability`:  with `tracked_field` / `untracked_field`
+        // - `revisions`:  with `tracked_field::maybe_changed_after`
+        // - `created_at`: with `untracked_field` / `maybe_changed_after`
```

Comment:
These I still need to check as to why we can say they are not going to end up aliasing (as Niko was talking about). I believe that is correct for all of them except for `revisions`, as that one is used to check if a field is outdated after all. Correct me if I am wrong but we first check that field to tell if we need to update the struct no? In which case we are in fact aliasing on that field